client/homebrew/editor/editor.jsx

```diff
@@ -102,6 +104,50 @@ const Editor = createClass({
 		}, 1);
 	},
 
+	addEditorPageNumbers : function(){
+		if(!this.refs.codeEditor) return;
+		if(this.state.view === 'text')  {
+			const codeMirror = this.refs.codeEditor.codeMirror;
+
+			const pageCountWidgets = document.getElementsByClassName('editor-page-count');
+			for (let x=pageCountWidgets.length - 1;x>=0;x--) pageCountWidgets[x].remove();
```

Comment:
I asked in the CodeMirror forums here https://discuss.codemirror.net/t/best-way-to-update-widgets-live-while-typing/3691 if there's some better way to do this than scan-and-delete with each keystroke the user types. This seems to work fine and it's basically what I did in the highlightCustomSyntax, but perhaps this is an opportunity to find a more efficient way to do this.

Reply:
I also thought this seemed a bit 'overkill'.  

When widgets are created, they are created not on the actual lines, but as a pile down at the bottom of a common container div (and then positioned with 'absolute' over the editor).  So possibly the 'search' can be focused on just that div rather than the whole document for a little extra efficiency?

Reply:
Hm... Well I have a couple responses from the CodeMirror dev here: https://discuss.codemirror.net/t/best-way-to-update-widgets-live-while-typing/3691/2

Doesn't look like there's a built-in thing, but possibly we can use some additional checks to only look at a particular range of lines. But then we need to track which widget belongs to which line so we only update the ones that are affected. And I suppose we also need to update any widgets *below* that line too, since the page count needs to be cascade down if any pages are added/removed above.

**Edit:**  Ooh, what about this? If we only worry about adding widgets if they are actually in the current scroll view? https://discuss.codemirror.net/t/get-visible-lines-with-respect-to-current-scroll/1429

We can still delete "all widgets" with each pass (to clear any offscreen widgets), but the total number of widgets we will ever be creating or destroying at a time will be limited to how many `\page` s can fit into the visible text.

Reply:
> do we want to include the "add pageLine class" into the "check if inside the current viewport" process?

Yes if it behaves properly. The one think I'm not sure of is scrolling up and down to see more of the document, if that will trigger redraws or if it only updates while typing.

Reply:
```js
const viewportRect = codeMirror.getWrapperElement().getBoundingClientRect();
const topVisibleLine = codeMirror.lineAtHeight(viewportRect.top, "window");
const bottomVisibleLine = codeMirror.lineAtHeight(viewportRect.bottom, "window");

_.forEach(this.props.brew.text.split('\n'), (line, lineNumber)=>{
if(this.props.renderer == 'legacy') {
	if(line.includes('\\page')) {

		// add a check here to see if in current viewport, then create widget if true
		if(lineNumber > topVisibleLine && lineNumber < bottomVisibleLine){
			const testElement = Object.assign(document.createElement('div'), {
				className   : 'editor-page-count',
				textContent : editorPageCount
			});
			codeMirror.addWidget({ line: lineNumber, ch: 0 }, testElement);
			testElement.style.top = `${parseInt(testElement.style.top) - codeMirror.defaultTextHeight()}px`;
			testElement.style.left = null;
		}
		
		// end createWidget process

		editorPageCount = editorPageCount + 1;
	}
}
```

So this is what I have.  It "works", except as you predicted it only updates on input, not on scroll.  I guess this is the same with highlightCustomMarkdown() as well.  

I haven't committed this yet to avoid a revert if possible.  Let me know if you have thoughts on how to make it update on scroll, as this is now beyond my current knowledge of JS/react.

Reply:
I was going to say, there's a line of code in CodeEditor.jsx that says CodeMirror.on("change") or something that fires the "redraw highlights" function. We could also add a similar line for "scroll". But "viewportchange" is probably better because it would cover resizing the window as well.

Reply:
@calculuschild I have experimented with CodeEditor.jsx and that onChange() but I'm stuck.  I can keep 'poking around' but it's kind of like a monkey at a typewriter at this point.

I can commit & push what I wrote above if you want, and let you (or someone else) figure out that last piece, or I can make some notes for myself and come back to it later when I think I have a firmer idea on it.

Reply:
@Gazook89 

Did you try adding a new line next to the `on("change")` one but using `on("viewportChange")` ?

```
this.codeMirror.on('viewPortChange', (cm)=>{this.props.onChange(cm.getValue());});
```

Reply:
Ah, didn't think it was that simple.  

So, that seems to work, but....scrolling takes a performance hit; it becomes kind of choppy.  I went ahead and utilized a new-to-me concept of "debouncing", and so it looks like this now

```js
this.codeMirror.on('viewportChange', _.debounce((cm)=>{this.props.onChange(cm.getValue());},200));
```

So there is a slight delay after scrolling, but improves the performance dramatically.

If this works for you, and there are no other changes needed, I'll go ahead and start to merge this function in with highlightCustomMarkdown() assuming we still want to go that route.

Reply:
Cool! Go ahead and merge it over.